Add iterators over hypercube vertices and edges

nilsvu opened the pull request:
## Proposed changes

These iterators help looping over all corners and edges of a cube. I'm using them to construct weighting functions for my Schwarz subdomain solver.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add iterators over hypercube elements

Utility to iterate over vertices, edges and higher-dimensional
elements (faces, cells, ...) on the boundary of a hypercube.

moxcodes commented on src/Domain/Structure/VertexIterator.hpp:
```diff
@@ -0,0 +1,31 @@
+
+/*!
+ * \brief Iterate over vertices in a `Dim`-dimensional hypercube
+ */
```
I feel these structures  should have significantly more description about their intended use and member functions.

How do you plan to use these? Most iterators in the STL and in spectre have either dereference operators, implicit conversion operators, or member functions to output the full state information so that can be passed easily into functions or objects that depend on that state. These currently have the appearance of requiring `Dim` calls to `side` to get complete state information, which seems cumbersome to me intuitively, but perhaps I'm not properly understanding your use case.

nilsvu replied:
Thanks for reviewing @moxcodes! Here's some "pseudocode" that illustrates how I currently use these iterators in the Schwarz weighting functions:

```cpp
// Computing weights on subdomains that overlap across corners and edges
for (VertexIterator<2> corner{}; corner; ++corner) {
  // Skip corners to external boundaries
  if (skip_corner(corner)) {
    continue;
  }
  for (size_t d = 0; d < 2; ++d) {
    get(corner_weight) *= intruding_weight(
        logical_coords.get(d), overlap_widths[d], corner.side(d));
  }
  // ...
}
for (EdgeIterator<2> edge{}; edge; ++edge) {
  const size_t d = edge.dimension();
  const Side side = edge.side(0);
  // Skip edges to external boundaries
  if (skip_edge(edge)) {
    continue;
  }
  get(edge_weight) *= intruding_weight(
      logical_coords.get(d), overlap_widths[d], side);
  // ...
}
```

These iterators are similar to the `SliceIterator` and `StripeIterator`. They encode the current state in the iterator object, which can be passed around (e.g. to the `skip_corner` and `skip_edge` functions).

Do you think these hypercube iterators can be useful, or should I inline their logic where I iterate over corners or edges in the Schwarz code? I'm not very attached to this code, but thought it could be a nice re-usable way of abstracting away the hypercube geometry. I'm happy to hear your suggestions!

moxcodes replied:
I do think that this sort of iterator can be useful -- from the code snippet, the only improvement that I'd suggest is try to construct the iterator in such a fashion that it could be used with range-based iteration (perhaps with some type of begin/end factory functions for the iterator?).

The edge part of this code seems counter-intuitive to me -- it has the appearance that if I have a consistent way of ordering the sides outside of the edge iterator, the edge iterator means different things as it advances (because it computes the upper/lower via a bitset of length dim-1) -- so as the iterator advances, if I'm interested in a particular side index, I will need to decrement it or omit the query to `edge.side` depending on what `edge.dimension` is.
Would it make the interface more difficult for your use-case if the `edge.side` function took the full range of dimensions and returned a `boost::optional`, giving `boost::none` if the side requested is invalid for the current edge?

nilsvu commented:
@moxcodes I addressed your feedback and ended up re-writing these iterators as a single class template that works for any k-dimensional element on the boundary of an n-dimensional hypercube. The iterator now has an `operator*` that returns a state-encoding vertex/edge/... as you suggested, and it works with range-based loops. To address your other concern I added a `side_in_parent_dimension(size_t)` function that works with the dimension of the hypercube, instead of the dimension of the edge(/face/...). Could you take another look and let me know what you think?

nilsvu commented:
@wthrowe could you perhaps do the second review?

moxcodes commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+
+  HypercubeElement<ElementDim, HypercubeDim> operator*() const noexcept;
+
+  bool operator!=(const HypercubeElementsIterator& rhs) const noexcept;
```
probably want `operator==` for the iterator too.

nilsvu replied:
It's not required for the an iterator. I would implement it as `return not(*this != rhs)`, which is ugly enough that I didn't do it. Do you think I should add it nonetheless?

moxcodes replied:
I agree it's not strictly required for the standard iterator use cases, and the standard seems slightly unclear, but I think in terms of interface design it's best to avoid the 'surprising' choices of defining either `operator==` or `operator!=` without the complimentary operator.

moxcodes commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+
+  HypercubeElementsIterator end() const noexcept;
+
+  const HypercubeElementsIterator& operator++() noexcept;
```
It's generally nice to define the prefix operator as well (denoted as `operator++(int)`), and that one can return by non-const reference

wthrowe replied:
To be (hopefully) a bit clearer here, the prefix operator (`operator++()`) should return a non-const reference, and a postfix operator (`operator++(int)`) should return by value.

moxcodes commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+           factorial(ElementDim) / factorial(num_indices);
+  }
+
+  HypercubeElementsIterator begin() const noexcept;
```
It's a little unconventional to have `begin` in the iterator class itself, I think, but I think it does make some sense for this case. But can the `begin()` and `end()` be static as they act as 'factory' methods for a fresh iterator?

moxcodes reviewed (commented):
Looks good, I like the generic utility. I just have a couple of tweaks to suggest for the iterator interface.

wthrowe commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+  template <size_t LocalElementDim = ElementDim,
+            Requires<LocalElementDim == 1> = nullptr>
+  HypercubeElement(size_t dim_in_parent,
+                   std::array<Side, HypercubeDim - 1> index) noexcept;
```
You have a lot of SFINAE constructor overloads here, and I think most of them are just saving you from having to enclose your arguments in double braces.  Is there a reason you need them?

nilsvu replied:
No reason other than convenience and a nicer interface. Take an `Edge` for example. It has only one `dim_in_parent`, no matter the dimension of the hypercube it's embedded in, so wrapping the `dim_in_parent` in double braces felt silly. The other constructor I could remove is `Vertex(Indices... indices)`. It just eliminates the double braces around the indices. Would you prefer I remove either, or both?

wthrowe replied:
I would drop both, but I won't object if you leave them unless someone else complains.

wthrowe commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+  /// `HypercubeDim`-dimensional hypercube.
+  static constexpr size_t size() noexcept {
+    return two_to_the(num_indices) * factorial(HypercubeDim) /
+           factorial(ElementDim) / factorial(num_indices);
```
If you intend to use this for high dimensions, this should be written in a way less susceptible to overflow.  My guess is you don't expect to use it that way, in which case this is fine.

nilsvu replied:
I'm only instantiating the template for `HypercubeDim<=3`, that's probably all we need.

wthrowe commented on tests/Unit/Domain/Structure/Test_Hypercube.cpp:
```diff
@@ -0,0 +1,146 @@
+    test_hypercube_iterator<2, 1, 4>({{Edge<2>{0, {{Side::Lower}}},
+                                       Edge<2>{0, {{Side::Upper}}},
+                                       {1, {{Side::Lower}}},
+                                       {1, {{Side::Upper}}}}});
```
`Edge<2>{...` for consistency.

wthrowe commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+
+  template <size_t LocalElementDim = ElementDim,
+            Requires<LocalElementDim == 0> = nullptr>
+  explicit HypercubeElement(std::array<Side, HypercubeDim> index) noexcept;
```
Test this.

wthrowe commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+
+  // @{
+  /// The parent hypercube's dimensions that this element shares
+  const std::array<size_t, ElementDim>& dimensions_in_parent() const noexcept;
```
Test this.

wthrowe commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+
+  template <size_t LocalElementDim = ElementDim,
+            Requires<LocalElementDim == 1> = nullptr>
+  size_t dimension_in_parent() const noexcept;
```
Test this.

wthrowe commented on src/Domain/Structure/Hypercube.hpp:
```diff
@@ -0,0 +1,172 @@
+  // @{
+  /// Whether this element is located on the lower or upper side in those
+  /// dimensions that it does not share with its parent hypercube
+  const std::array<Side, HypercubeDim - ElementDim>& index() const noexcept;
```
Test this.

And so on.  Make sure everything's tested.

nilsvu commented:
@moxcodes @wthrowe thanks for your reviews, fixups posted!

moxcodes commented:
I'm happy for you to squash (along with the suggested equality operator), pending @wthrowe 's go-ahead.

nilsvu commented:
@wthrowe Could you check the fixups please?

wthrowe commented on src/Domain/Structure/Hypercube.cpp:
```diff
@@ -200,6 +205,15 @@ bool HypercubeElementsIterator<ElementDim, HypercubeDim>::operator!=(
+    const HypercubeElementsIterator& rhs) const noexcept {
+  // Implemented in terms of `operator!=` because an iterator only requires the
+  // inequality operator, and this equality operator is implemented only to
+  // complete the interface
```
I don't understand this comment.  ("An iterator only requires the inequality operator"?  Input iterators certainly need both.)  Just do comparisons the same way they are done everywhere else.  Then you don't have to explain it.

Also, binary operators should be free functions unless they access private state.

nilsvu commented:
@wthrowe pushed a fixup for the `operator==`, could you take another look?

wthrowe commented:
`!=` should be a free function.  And actually I think we've been doing `==` as a friend function, rather than a member.

nilsvu commented:
@wthrowe right of course, fixed

wthrowe commented:
Squash.

moxcodes reviewed (approved):
Looks good to me

nilsvu commented:
Thanks for reviewing @moxcodes and @wthrowe! Seems ready to merge @wthrowe
